tsdb/engine/tsm1/engine.go

```diff
@@ -228,7 +258,8 @@ func (e *DevEngine) Begin(writable bool) (tsdb.Tx, error) {
 
 func (e *DevEngine) WriteTo(w io.Writer) (n int64, err error) { panic("not implemented") }
 
-func (e *DevEngine) writeSnapshot() error {
+// WriteSnapshot will snapshot the cache and write a new TSM file with its contents, releasing the snapshot when done.
+func (e *DevEngine) WriteSnapshot(async bool) error {
```

Comment:
There are multiple ways to handle all of those situations without crashing and still alert operators to problems that may need intervention.  If a database does panic, it should be for a very good reason.  Out of disk space is not a good reason IMO.  We should expect that to happen and design a solution that handles it gracefully.  Relying on your database crashing to alert you to the fact that you are not properly monitoring your disk usage does not make sense to me.